core/src/main/java/jenkins/security/auth/LoginThrottler.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.security.auth;
+
+import hudson.Extension;
+import jenkins.model.Jenkins;
+import jenkins.security.SecurityListener;
+import jenkins.util.TimeProvider;
+import org.acegisecurity.BadCredentialsException;
+import org.acegisecurity.userdetails.UserDetails;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+@Extension
+public class LoginThrottler extends SecurityListener {
+    private static final Logger LOGGER = Logger.getLogger(LoginThrottler.class.getName());
+
+    private static final Integer RATE_LIMITER_WINDOW_DURATION = Integer.getInteger(LoginThrottler.class.getName() + ".duration", 30);
+    private static final Integer RATE_LIMITER_MAX_ATTEMPTS = Integer.getInteger(LoginThrottler.class.getName() + ".attempts", 3);
+
+    private ReadWriteLock failuresLock = new ReentrantReadWriteLock();
+    private Map<String, UserLoginFailuresInfo> failures = new HashMap<>();
+    private boolean active = true;
+
+    private static class UserLoginFailuresInfo {
+        private SortedSet<Date> failures;
+
+        private String name;
+
+        private UserLoginFailuresInfo(String name) {
+            failures = new TreeSet<>();
+            this.name = name;
+        }
+
+        public void recordFailure() {
+            LOGGER.log(Level.FINE, () -> "Recording login failure for " + name);
```

Comment:
This would be better integrated with audit logging, than using the general logging mechanism. Same comment for a number of other log messages.